### PR TITLE
Improve performance of the frequency rule query

### PR DESCRIPTION
**Pull request by alanhartless**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This improves the frequency rule query by breaking up into two queries, one for those with custom rules and one without. For the default rules, it allows to do simple calculations in PHP (such as the date) rather than asking MySQL to calculate on the fly. 

Original (table with 133 million email stats): 32.8 seconds
```
SELECT ch.lead_id, count(ch.lead_id) 
FROM mautic_email_stats ch 
WHERE (ch.date_sent >= '2018-11-08 21:23:54') AND (ch.lead_id IN (...)) 
GROUP BY ch.lead_id, fr.frequency_time, fr.frequency_number 
HAVING (count(ch.lead_id) >= fr.frequency_number AND ch.lead_id IN (...))
```

Broken into two queries: total 528.7ms
with custom rules: 49.7 ms
```
SELECT ch.lead_id, fr.frequency_number, fr.frequency_time 
FROM mautic_email_stats ch 
INNER JOIN mautic_lead_frequencyrules fr ON ch.lead_id = fr.lead_id 
WHERE fr.channel = 'email' AND ((ch.date_sent >= case fr.frequency_time
                 when 'MONTH' then DATE_SUB(NOW(),INTERVAL 1 MONTH)
                 when 'DAY' then DATE_SUB(NOW(),INTERVAL 1 DAY)
                 when 'WEEK' then DATE_SUB(NOW(),INTERVAL 1 WEEK)
                end)) AND (ch.lead_id IN (...)) 
GROUP BY ch.lead_id, fr.frequency_time, fr.frequency_number 
HAVING count(ch.lead_id) >= fr.frequency_number;
```

default for those without custom rules: 479ms
```
SELECT ch.lead_id
FROM mautic_email_stats ch 
WHERE (ch.date_sent >= '2018-11-08 21:23:54') AND (ch.lead_id IN (...)) 
    AND (NOT EXISTS (SELECT null FROM mautic_lead_frequencyrules fr WHERE fr.lead_id = ch.lead_id)) 
GROUP BY ch.lead_id 
HAVING count(ch.lead_id) >= 3;
``` 

[//]: # ( As applicable: )
#### Steps to test this PR:
1. Define a default email frequency rule in Mautic's configuration on the email settings tab
2. Go to the preference center for a single contact and define a different custom frequency rule (allow more than the default)
3. Add the contact above along with some others into a segment
4. Send emails to these contacts until they hit the frequency rule for both and emails are starting to queue
5. The contact with a custom frequency rule should not have had emails queued until it hit it's limit
6. All other contacts should have started queuing emails once it hit the limit as defined in Mautic's configuration
7. Repeat the test without a global configuration option set and a new contact with custom frequency rules and newly added to the segment)
8. Send emails again and this time only the contact with a custom frequency rule should have mail queued. The rest should be sent since there is no default.

### CODE REVIEW

**Comment by jojomnky** [comment]
I was able to achieve a similar performance enhancement by simply reversing the from/join on the original query. Since there should be far fewer lead_frequencyrules than there are stats. 

eg. switching

```
SELECT ch.lead_id, fr.frequency_number, fr.frequency_time 
FROM mautic_email_stats ch 
INNER JOIN mautic_lead_frequencyrules fr ON ch.lead_id = fr.lead_id 
WHERE fr.channel = 'email' AND ((ch.date_sent >= case fr.frequency_time
                 when 'MONTH' then DATE_SUB(NOW(),INTERVAL 1 MONTH)
                 when 'DAY' then DATE_SUB(NOW(),INTERVAL 1 DAY)
                 when 'WEEK' then DATE_SUB(NOW(),INTERVAL 1 WEEK)
                end)) AND (ch.lead_id IN (...)) 
GROUP BY ch.lead_id, fr.frequency_time, fr.frequency_number
WHERE (ch.date_sent >= '2018-11-08 21:23:54') AND (ch.lead_id IN (...)) 
```
 to 

```
SELECT ch.lead_id, fr.frequency_number, fr.frequency_time 
FROM mautic_lead_frequencyrules fr
INNER JOIN  mautic_email_stats ch ON ch.lead_id = fr.lead_id 
WHERE .....
```

File: app/bundles/LeadBundle/Entity/FrequencyRuleRepository.php [hunk 1]
```diff
@@ -19,69 +19,46 @@
...
-
-        $havingAnd = 'AND '.$q->expr()->in("ch.$statContactColumn", $leadIds);
-        if ($defaultFrequencyNumber != null) {
-            $q->having("(count(ch.$statContactColumn) >= IFNULL(fr.frequency_number,:defaultNumber) $havingAnd)")
-                ->setParameter('defaultNumber', $defaultFrequencyNumber);
-        } else {
-            $q->having("(count(ch.$statContactColumn) >= fr.frequency_number $havingAnd)");
+        $violations = $this->getCustomFrequencyRuleViolations($channel, $leadIds, $statTable, $statContactColumn, $statSentColumn);
+
+        if ($defaultFrequencyTime && $defaultFrequencyTime) {
```
Comment (maxlawton) [inline app/bundles/LeadBundle/Entity/FrequencyRuleRepository.php:47]: Is this condition supposed to be `&&` of `$defaultFrequencyTime` and itself, or should it be `($defaultFrequencyNumber && $defaultFrequencyTime)` ?

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by RCheesley** [comment]
@alanhartless conflicts to be addressed please!

File: app/bundles/LeadBundle/Entity/FrequencyRuleRepository.php (hunk 1 above)
Comment (dongilbert) [reply app/bundles/LeadBundle/Entity/FrequencyRuleRepository.php:47]: Supposed to be both.

**Comment by dongilbert** [comment]
@RCheesley I've resolved conflicts. And yes, still relevant for M3

**Comment by RCheesley** [comment]
Awesome, many thanks!

**Comment by dennisameling** [comment]
@dongilbert Travis has failed again :( would you mind checking it? Thanks in advance!!

**Comment by RCheesley** [comment]
Bumping to 3.2 due to unresolved Travis failure

**Comment by RCheesley** [comment]
@cla-bot check

**Comment by RCheesley** [comment]
@alanhartless looks like this one is going to decrease coverage according to CodeCov so would need to improve test coverage I think.

**Review by kuzmany — DISMISSED** [review]
I've test few combinations with frequency rule and works as expected 👍 We should go to 3.3.0

**Comment by RCheesley** [comment]
@alanhartless @mautic/acquia-po this one needs rebasing and is on a protected branch so we cannot help with that :(

**Comment by anton-vlasenko** [comment]
I've added a functional test to increase code coverage, but it's failing here and it doesn't fail in my local environment.
This needs more investigation and I don't have time to do it now.
Possible code contributors: feel free to modify or delete it (app/bundles/LeadBundle/Tests/Entity/FrequencyRuleRepositoryTest.php)
I will get to it on my next "duty" if it isn't fixed by then.

**Comment by RCheesley** [comment]
@anton-vlasenko I have just pulled in the changes from the features branch - maybe that was causing the issue? Will see once the tests are done running!

**Comment by RCheesley** [comment]
@alanhartless @anton-vlasenko @fedys it looks like we still have a [failing test here](https://github.com/mautic/mautic/pull/7191/checks?check_run_id=2591601036#step:9:70). Could you take a look please? If we can get that passing we are good to merge assuming code coverage returns a ✅

**Comment by fedys** [comment]
@RCheesley I fixed the failing test.

**Comment by RCheesley** [comment]
@all-contributors please add @alanhartless for code

**Pull request merged**